compiler/rustc_mir/src/transform/separate_const_switch.rs

```diff
@@ -0,0 +1,279 @@
+//! A pass that copies switch-terminated blocks into
+//! another copy for each parent that set the value
+//! being switched over to a constant.
+
+use crate::transform::MirPass;
+use rustc_middle::mir::*;
+use rustc_middle::ty::TyCtxt;
+use smallvec::SmallVec;
+
+pub struct SeparateConstSwitch;
+
+impl<'tcx> MirPass<'tcx> for SeparateConstSwitch {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        if tcx.sess.mir_opt_level() < 4 {
+            return;
+        }
+
+        separate_const_switch(body);
+    }
+}
+
+pub fn separate_const_switch<'tcx>(body: &mut Body<'tcx>) {
+    let mut new_edges: SmallVec<[(BasicBlock, BasicBlock); 6]> = SmallVec::new();
+    let predecessors = body.predecessors();
+    'block_iter: for (block_id, block) in body.basic_blocks().iter_enumerated() {
+        if let TerminatorKind::SwitchInt {
+            discr: Operand::Copy(switch_place) | Operand::Move(switch_place),
+            ..
+        } = block.terminator().kind
+        {
+            // if the block is on an unwind path, do not
+            // apply the optimization as unwind paths
+            // rely on a unique parent invariant
+            if block.is_cleanup {
+                continue 'block_iter;
+            }
+
+            // if the block has fewer than 2 predecessors, ignore it
+            // we could maybe chain blocks that have exactly one
+            // predecessor, but for now we ignore that
+            if predecessors[block_id].len() < 2 {
+                continue 'block_iter;
+            }
+
+            // first, let's find a non-const place
+            // that determines the result of the switch
+            if let Some(switch_place) = find_determining_place(switch_place, block) {
+                // we now have an input place for which it would
+                // be interesting if predecessors assigned it from a const
+
+                let mut predecessors_left = predecessors[block_id].len();
+                'predec_iter: for predecessor_id in predecessors[block_id].iter().copied() {
+                    if let Some(predecessor) = body.basic_blocks().get(predecessor_id) {
+                        // first we make sure the predecessor jumps
+                        // in a reasonable way
+                        match &predecessor.terminator().kind {
+                            // the following terminators are
+                            // unconditionally valid
+                            TerminatorKind::Goto { .. } | TerminatorKind::SwitchInt { .. } => {}
+
+                            TerminatorKind::FalseEdge { real_target, .. } => {
```

Comment:
No idea! If they are then they have no reason to be handled.

Reply:
After checking, apparently it is the job of SimplifyBranch to convert them. So as it technically can be ran on code that still has them, I think it's preferable to handle them, just in case.